### ISSUE TITLE
make pre-commit persisting errors

### ISSUE THREAD

**Issue by emilyriehl** [UniMath/agda-unimath#633]
make pre-commit gives the following output, even on repeated runs:

```
Trim Trailing Whitespace.....................................................Passed
Fix End of Files.............................................................Passed
Mixed line ending............................................................Passed
Fix double quoted strings....................................................Passed
Check python ast.............................................................Passed
Check Yaml...................................................................Passed
Check for case conflicts.....................................................Passed
Check for merge conflicts....................................................Passed
Check for added large files..................................................Passed
Markdown conventions.........................................................Failed
- hook id: markdown-conventions
- exit code: 1

Traceback (most recent call last):
  File "scripts/markdown_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/markdown_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/markdown_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/markdown_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/markdown_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/markdown_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/markdown_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/markdown_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable

Blank line conventions in literate Agda Markdown files.......................Failed
- hook id: blank-line-conventions
- exit code: 1

Traceback (most recent call last):
  File "scripts/blank_line_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/blank_line_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/blank_line_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/blank_line_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/blank_line_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/blank_line_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/blank_line_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/blank_line_conventions.py", line 6, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable

Format Agda imports..........................................................Failed
- hook id: fix-imports
- exit code: 1

Traceback (most recent call last):
  File "scripts/fix_imports.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/fix_imports.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/fix_imports.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/fix_imports.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/fix_imports.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/fix_imports.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/fix_imports.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/fix_imports.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable

Simple Agda space conventions................................................Failed
- hook id: spaces-conventions-simple
- exit code: 1

Traceback (most recent call last):
  File "scripts/spaces_conventions_simple.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/spaces_conventions_simple.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/spaces_conventions_simple.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/spaces_conventions_simple.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/spaces_conventions_simple.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/spaces_conventions_simple.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/spaces_conventions_simple.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/spaces_conventions_simple.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable

Fix simply wrappable long lines of Agda code.................................Failed
- hook id: simply-wrappable-long-lines
- exit code: 1

Traceback (most recent call last):
  File "scripts/simply_wrappable_long_lines.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/simply_wrappable_long_lines.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/simply_wrappable_long_lines.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/simply_wrappable_long_lines.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/simply_wrappable_long_lines.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/simply_wrappable_long_lines.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/simply_wrappable_long_lines.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable
Traceback (most recent call last):
  File "scripts/simply_wrappable_long_lines.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable

Check the maximum line length conventions....................................Failed
- hook id: max-line-length-conventions
- exit code: 1

Traceback (most recent call last):
  File "scripts/max_line_length_conventions.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable

Generate Agda namespace index modules........................................Failed
- hook id: generate-namespace-index-modules
- exit code: 1

Traceback (most recent call last):
  File "scripts/generate_namespace_index_modules.py", line 8, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable

Generate main index file.....................................................Failed
- hook id: generate-main-index-file
- exit code: 1

Traceback (most recent call last):
  File "scripts/generate_main_index_file.py", line 7, in <module>
    import utils
  File "/home/eriehl/Math/Hopkins/Formalization/agda-unimath/scripts/utils/__init__.py", line 21, in <module>
    def find_index(s: str, t: str) -> list[int]:
TypeError: 'type' object is not subscriptable

Python scripts formatting....................................................Passed
CSS, JS, YAML and Markdown (no codeblocks) formatting........................Passed
make: *** [Makefile:90: pre-commit] Error 1
```

**Comment by VojtechStep** [comment]
I think that might be a Python version issue - the `list[int]` generic syntax, which the error complains about, should be available since Python 3.9; which version are you using?

**Comment by emilyriehl** [comment]
I'm using ubuntu in case this is relevant. 

When I type
```
python --version
```
it tells me the command is not found. When I type
```
python3 --version
```
it returns Python 3.8.10. I've just now managed to install python3.9 so

```
python3.9 --version
```
returns Python 3.9.16 (but the python3 version is unchanged).

I believe I may be using a package called "python-is-python3". I forget exactly why. The make pre-commit error message looks the same as before.

**Comment by VojtechStep** [comment]
I personally don't have experience with how Ubuntu handles Python updates, so take the following with a grain of salt.

If running `python` says that the command is not found, then that's perfectly fine - a Python script should specify whether it's using `python3` or `python2`, and having the `python` command as a shorthand for one of them is more of a user-facing convenience. That's actually what the `python-is-python3` package does - it creates a link so that running just `python` is the same as running `python3` (and since `python` isn't found on your system, you don't have `python-is-python3` installed).

Since you installed the `python3.9` package, you now have two python installations on your system - the default Python 3.8 (accessible via `python3`), and the new Python 3.9 (accessible via `python3.9`). If you want to make Python 3.9 the default (thus a script beginning with `#!/usr/bin/env python3`, such as our pre-commit hooks, would run that instead of 3.8), then you can use Ubuntu's `update-alternatives` tool.

However, as I'm reading more into this, it seems that people usually caution against messing with the default `python3` version, since pretty much all system components have a dependency on it, and who knows what changing the version would break.

I think the better solution would be for our pre-commit hooks to support Python 3.8, since that's the default for Ubuntu 20.04 LTS, which will be supported until mid-2024 (with "Extended Security Maintanance" period reaching into 2030). I'll look into it.

**Comment by emilyriehl** [comment]
Thanks! I'll wait for further instructions.

**Comment by VojtechStep** [comment]
Ok, so it seems like pre-commit actually uses its own python, not the system-wide version, which is good. But unfortunately I can't reproduce your issue, my hooks run just fine.

On my system, there's this directory `~/.cache/pre-commit/repoabbobbu8/py_env-python3.10/`. Could you check if you have something similar? I'm not sure if the `repoabbobbu8` part is deterministic, and there might be multiple such folders under `~/.cache/pre-commit`, but could you look around and find one with a `py_env-python-3.X` subdirectory?

**Comment by VojtechStep** [comment]
Well, the Python version selection story is a little more disappointing than I thought, so I went and changed the offending piece of code to be compatible with Python 3.8; this PR should fix the issue: #634.
